mne/preprocessing/artifact_detection.py

```diff
@@ -0,0 +1,316 @@
+
+    Data is band pass filtered at high frequencies, smoothed with the envelope,
+    z-scored, channel averaged and low-pass filtered to remove transient peaks.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The data.
+    thr : integer
+        threshold where a segment is marked as artifactual. The thr represent
+        the channel averaged z-score values of the data band pass filtered
+        betweehn 110 and 140 Hz.
+    t_min = integer
+        Min. time between annotated muscle artifacts, below will be annotated
+        as muscle artifact.
+    notch = List
+        The frequencies at which to apply a notch filter
+
+    Returns
+    -------
+    annot : mne.Annotations
+        periods with muscle artifacts
+    art_scores_filt : array
+        the z-score values of the filtered data
+    """
+    raw.pick_types(meg=True, ref_meg=False)
+    if notch is not None:
+        raw.notch_filter(notch, fir_design='firwin')
+    raw.filter(110, 140, fir_design='firwin')
+    raw.apply_hilbert(envelope=True)
+    sfreq = raw.info['sfreq']
+    art_scores = zscore(raw._data, axis=1)
+    # band pass filter the data
+    art_scores_filt = filter_data(art_scores.mean(axis=0), sfreq, None, 4)
+    art_mask = art_scores_filt > thr
+    # remove artifact free periods shorter than t_min
+    idx_min = t_min * sfreq
+    comps, num_comps = label(art_mask == 0)
+    for l in range(1, num_comps + 1):
+        l_idx = np.nonzero(comps == l)[0]
+        if len(l_idx) < idx_min:
+            art_mask[l_idx] = True
+    mus_annot = _annotations_from_mask(raw.times, art_mask,
+                                       'Bad-muscle')
+    return mus_annot, art_scores_filt
+
+
+def weighted_median(data, weights):
+    """Get median with weights assigned to the data.
+
+    Parameters
+    ----------
+    data : array
+    weights : array or list
+        values to assign at each of the data points
+    """
+    dims = data.shape
+    w_median = np.zeros((dims[1], dims[2]))
+    for d1 in range(dims[1]):
+        for d2 in range(dims[2]):
+            data_dd = np.array(data[:, d1, d2]).squeeze()
+            s_data, s_weights = map(np.array, zip(*sorted(zip(data_dd,
+                                                              weights))))
+            midpoint = 0.5 * sum(s_weights)
+            if any(s_weights > midpoint):
+                w_median[d1, d2] = (data[weights == np.max(weights)])[0]
+            else:
+                cs_weights = np.cumsum(s_weights)
+                idx = np.where(cs_weights <= midpoint)[0][-1]
+                if cs_weights[idx] == midpoint:
+                    w_median[d1, d2] = np.mean(s_data[idx:idx + 2])
+                else:
+                    w_median[d1, d2] = s_data[idx + 1]
+    return w_median
+
+
+def _annotations_from_mask(times, art_mask, art_name):
+    """Make annotations."""
+    comps, num_comps = label(art_mask)
+    onsets = []
```

Comment:
maybe `onsets, durations, desc = [], [], []` (or `... = list(), list(), list()` as it is sometimes preferred in mne)